src/SessionSetup/ConnectToTargetDialog.cpp

```diff
@@ -31,160 +32,143 @@ ConnectToTargetDialog::ConnectToTargetDialog(
+                       } else {
+                         LogAndDisplayError(QString::fromStdString(result.error().message()));
+                         reject();
+                       }
+                     });
+
+  int rc = QDialog::exec();
+
+  if (rc != QDialog::Accepted) {
+    return std::nullopt;
   }
 
-  CloseDialogIfResultIsReady();
+  return target;
 }
 
-void ConnectToTargetDialog::LoadInstanceDetailsAsync() {
-  ggp_client_->DescribeInstanceAsync(instance_id_)
-      .Then(main_thread_executor_.get(),
-            [this](ErrorMessageOr<orbit_ggp::Instance> instance_result) {
-              if (instance_result.has_error()) {
-                result_data_.maybe_instance_ = std::nullopt;
-                result_data_.error_message_ =
-                    QString::fromStdString(instance_result.error().message());
-              } else {
-                result_data_.maybe_instance_ = instance_result.value();
-              }
-              CloseDialogIfResultIsReady();
-            });
-}
+ErrorMessageOr<StadiaTarget> ConnectToTargetDialog::OnAsyncDataAvailable(
+    MaybeSshAndInstanceData ssh_instance_data) {
+  auto& [ssh_info_result, instance_result] = ssh_instance_data;
+
+  std::string error;
+  if (ssh_info_result.has_error()) {
+    return ssh_info_result.error();
+  } else if (instance_result.has_error()) {
+    return instance_result.error();
+  }
+
+  ConnectionData connection_data;
 
-std::optional<ServiceDeployManager::GrpcPort> ConnectToTargetDialog::DeployOrbitService(
-    orbit_ggp::SshInfo& ssh_info) {
-  result_data_.service_deploy_manager_ =
+  connection_data.service_deploy_manager_ =
       std::make_unique<orbit_session_setup::ServiceDeployManager>(
           ssh_connection_artifacts_->GetDeploymentConfiguration(),
-          ssh_connection_artifacts_->GetSshContext(), CredentialsFromSshInfo(ssh_info),
+          ssh_connection_artifacts_->GetSshContext(),
+          CredentialsFromSshInfo(ssh_info_result.value()),
           ssh_connection_artifacts_->GetGrpcPort());
 
-  orbit_ssh_qt::ScopedConnection label_connection{QObject::connect(
-      result_data_.service_deploy_manager_.get(), &ServiceDeployManager::statusMessage, this,
-      &ConnectToTargetDialog::SetStatusMessage)};
+  OUTCOME_TRY(auto&& grpc_port, DeployOrbitService(connection_data.service_deploy_manager_.get()));
+  connection_data.grpc_channel_ = CreateGrpcChannel(grpc_port.grpc_port);
+
+  OUTCOME_TRY(connection_data.process_data_,
+              FindSpecifiedProcess(connection_data.grpc_channel_, process_id_));
+
+  return CreateTarget(std::move(connection_data), std::move(instance_result.value()));
+}
+
+StadiaTarget ConnectToTargetDialog::CreateTarget(ConnectionData result,
+                                                 orbit_ggp::Instance instance) const {
+  CHECK(instance.id == instance_id_);
+  CHECK(result.grpc_channel_ != nullptr);
+  CHECK(result.service_deploy_manager_ != nullptr);
+  CHECK(result.process_data_ != nullptr);
```

Comment:
Again pers pref, but I'd get rid of this (the whole CreateTarget function) and put the necessary code directly into OnAsyncDataAvailable

Reply:
Even though the checks are somewhat unneeded here, I'd like to leave this as a separate method. I prefer having smaller methods with descriptive names to make the overall codeflow more readable without needing to dive into any details, even if it results in slightly more code.